### PR TITLE
Updated id system

### PR DESCRIPTION
**Pull request by CodingRays**
Documentation should say everything.

UUIDs are now 128bit values which are split into a global and local part.
The global part is to ensure global uniqueness while the local part allows systems to generate ids that are efficient for it or contain additional metadata.

The NamedUUID has been updated to use a special global id while using the previously used hash for the local id.

Both global and local ids contain niches for efficient use of options.

### CODE REVIEW

**Comment by CodingRays** [comment]
Wait nevermind this split local id is stupid. Im gonna change it to a NonZeroU64.

**Comment by CodingRays** [comment]
Much simpler and more flexible now

File: src/util/id.rs [hunk 1]
```diff
@@ -1,33 +1,311 @@
...
+}
+
+/// A UUID generated from a string.
+///
+/// NamedUUIDs use a predefined global id with the local id being calculated as the hash of a
+/// string. The name is stored along side the UUID for easy debugging or printing. The name is
+/// stored by Arc enabling fast Copying of the struct.
+#[derive(Clone, Debug)]
+pub struct NamedUUID {
+    name: Arc<String>,
```
Comment (ramidzkh) [inline src/util/id.rs:228]: Wouldn't an Arc have more overhead than a String clone?

File: src/util/id.rs [hunk 2]
```diff
@@ -1,33 +1,311 @@
...
+///
+/// // 0 is a niche so Options are free
+/// assert_eq!(8, std::mem::size_of::<Option<GlobalId>>());
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct GlobalId(NonZeroU64);
+
+impl GlobalId {
+    // Need to reserve value 1 for the NamedId address space
+    const NEXT_GLOBAL_ID: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(2u64);
```
Comment (ramidzkh) [inline src/util/id.rs]: `use` it?

File: src/util/id.rs [hunk 3]
```diff
@@ -1,33 +1,311 @@
...
+///
+/// // Creates a new different global id
+/// let other_id = GlobalId::new();
+///
+/// assert_ne!(id, other_id);
+///
+/// // 0 is a niche so Options are free
+/// assert_eq!(8, std::mem::size_of::<Option<GlobalId>>());
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
```
Comment (ramidzkh) [inline src/util/id.rs:41]: What does it mean to be Ord for an id? I guess for BTreeMaps actually

**Review by ramidzkh — COMMENTED** [review]
I feel that two 64 bit counters seem excessive

File: src/util/id.rs (hunk 1 above)
Comment (CodingRays) [reply src/util/id.rs:228]: A string clone needs to allocate new memory and copy the entire string.
A arc clone is incrementing a atomic int

File: src/util/id.rs (hunk 2 above)
Comment (CodingRays) [reply src/util/id.rs]: This is the only place where AtomicU64 is needed so a use seemed excessive. But ill agree that it looks odd... should probably change it

File: src/util/id.rs (hunk 3 above)
Comment (CodingRays) [reply src/util/id.rs:41]: Yeah its needed for some data structures and algorithms where it doesnt matter what kind of order you have but you just need a order. Its just the natural order over the underlying payload.

**Comment by CodingRays** [comment]
As for the 2 64 bit value. These ids may be allocated very frequently. So a 32 bit type would run out too quickly. But doing a 48bit 16bit split wouldnt give enough address space to local ids. 96 bits or any other value between 64 and 128 doesnt really make any sense since the free bytes wont get used by other stuff anyways most of the time. 2 64Bit values really arent expensive at all and for the extra flexibility it adds i felt it was completely justified.

**Review by ramidzkh — APPROVED** [review]

**Comment by CodingRays** [comment]
Will be merged together with the init system rework. There have been some critical bug fixes added there.

**Pull request closed without merging**